### DISCUSSION TITLE
PKCE support for OAuth (required as of OAuth 2.1)

### DISCUSSION THREAD

**Discussion by michaelcbrook** [meteor/meteor#11759] [Feature Requests]
Hey there,

I notice no one has brought this up yet, but more and more service providers are starting to require PKCE as part of the OAuth authorization code flow in anticipation of this standard becoming required in [OAuth 2.1](https://oauth.net/2.1/). But this does not currently seem to be supported by Meteor.

[Read more about PKCE](https://oauth.net/2/pkce/)

[Simple PKCE example](https://github.com/curityio/pkce-javascript-example)

Are there any plans to offer support for this going forward?

---

The use case here is that I'm creating an accounts package to work with a service provider that requires PKCE. The ideal flow should work as follows:

1. The accounts package creates a random secret (called the `code_verifier`)
2. The accounts package transforms the `code_verifier` to produce a `code_challenge` (usually via SHA-256 hashing)
3. The `code_challenge` is sent in the initial request for the authorization code, along with `code_challenge_method=S256`
4. After the authorization code is received, the server makes the second request to exchange the code for an access token, where it additionally sends the `code_verifier` from step 1 along with the request.
5. If the authorization server approves the request, an access token is returned.

In my package, I am able to handle steps 1-3, but I am unable to see a way to get the `code_verifier` from step 1 to the server to send in step 4 without exposing the `code_verifier` to the authorization server by sending it via the `state` param.

Perhaps someone with more insight into the inner workings of the oauth package (or related packages) can shed some light on whether there is a way to achieve this or whether an improvement needs to be made to the oauth package in order to support PKCE flows.

Thank you!

**Comment by StorytellerCZ** [comment]
I'm looking into something similar with the existing OAuth to increase security where an additional secret is generated. Right now I'm kind of split if to make adjustments to the accounts system that would allow to store this secret for a limited time in the DB or if a better approach might be to use a same mechanism to generate the secret on the server as well in exchange for it not being totally random.

**Comment by michaelcbrook** [comment]
PKCE is specifically what I'm interested in being added to Meteor. Even though they're looking to make it required in OAuth 2.1, PKCE itself is still very much part of OAuth 2.0. It's also not exclusive to public clients. PKCE covers [additional attack vectors](https://security.stackexchange.com/questions/219105/pkce-vs-client-secret) even for confidential clients that a `client_secret` still does not cover.

I think with the world headed towards requiring PKCE in OAuth in the not-too-distant future, this is something meteor should stay on top of before meteor becomes incompatible with newer OAuth servers. It could become a deal killer for developers looking to use meteor at some point.

@StorytellerCZ I don't completely follow, but I don't think storing an additional secret in the database would solve this problem because a random secret would need to be generated for each login. In the `-oauth` packages, I notice there's a `credentialToken` which is a random secret that identifies each login, but unfortunately, this is sent to the identity provider in the `state` param in a base64 encoded string. So it's technically readable in the URL to the identity provider, making it not very safe to be the `code_verifier`. I believe a second secret would need to be created. Are you perhaps talking about storing the new secret with the `credentialToken` in the database for each login? (I haven't checked, but I assume `credentialToken` gets stored there)

**Reply by StorytellerCZ** [reply]
Possibly. I think first step here would be to identify providers that have PKCE implemented already so that we can look on their specific implementations and start toying with things. If I remember correctly GitHub and Facebook might have something like this, but I don't think they call it PKCE.
Do you have any information on specific providers?

**Reply by StorytellerCZ** [reply]
Just a side note, the following might be useful for this in the future:
https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API

**Reply by michaelcbrook** [reply]
@StorytellerCZ I know I haven't touched this forum in several months, but we're in our next round of audits, and this issue came up again. Doing a quick Google search, here are a few services that have implemented PKCE:

- Dropbox: https://dropbox.tech/developers/pkce--what-and-why-
- Microsoft Identity Platform: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow
- PagerDuty: https://developer.pagerduty.com/docs/ZG9jOjExMDI5NTc1-o-auth-2-0-pkce-flow

And going forward, our company [Pitchly](https://www.pitchly.com/), is also planning on making PKCE available (and requiring it) as a service provider, but this hasn't been documented yet.

In Microsoft's documentation, it reads:

> [The code_challenge] parameter is now recommended for all application types, both public and confidential clients, and required by the Microsoft identity platform for [single page apps using the authorization code flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/reference-third-party-cookies-spas).

> PKCE is required for SPAs on the Microsoft identity platform. PKCE is recommended for native and confidential clients.

So Microsoft now **REQUIRES** it for all Single-page apps using the authorization code flow on the Identity Platform.

I still have not been able to find a straightforward way to implement this in Meteor, but PKCE is still slated to become required by [OAuth 2.1](https://oauth.net/2.1/).

**Reply by StorytellerCZ** [reply]
I guess I will have to bump this on my priority list. Though I'm still not entirely sure that I completely understand this.
This is how I see it so far:

1. When the OAuth 2.1 flow is triggered, Meteor will generate `code_verifier` and `code_challenge`. `code_verifier` is generated via `crypto.randomBytes` or the Random package and base64 encoded. `code_challenge` is `code_verifier` sha256 encrypted and once again base64 encoded.
2. `code_challenge` is send to the request endpoint together with `code_challenge_method` set to `s256` (sha256).
3. `code_verifier` is saved in the DB, probably will have to include `code_challenge` as well, so that we can use it to later find it (step 5), or we can store it in `localStorage`. I think `localStorage` makes a bit more sense from security standpoints, but I'm very much open on guidance here.
4. User authorizes the service and the service sends `authorization_code` to the `redirect_uri`.
5. Meteor will grab `authorization_code` and `code_verifier` (earlier stored in DB or `localStorage`) and request for `token`
6. If `code_verifier` and `code_challenge` match then the service will return `token`
7. Cleanup & continue as usual

Did I get that right?

For this we will have to adjust OAuth2 package to allow for this flow. Probably just add an option when starting the flow that is going to be set by the service package.

**Reply by StorytellerCZ** [reply]
OK, started to do a bit of a digging in the code. I think we can add version `2.1` when registering a service and then in `oauth2_server` we add `OAuth._requestHandlers['2.1']` with the specifics for the 2.1 flow. I think this part is for receiving the data from step 5.
We will have to add the `code_challenge` and `code_challenge_method` to the login uris in the service package (step 2). This also looks like a good place to do the generations of the codes (step 1), or probably at the very beginning `OAuth.launchLogin`, tough to say now.
I will probably have to make a diagrams on how the internals work so that we can add this new workflow.

**Comment by jankapunkt** [comment]
@StorytellerCZ I would like to join on this one, I have at least some oauth background by collaborating on https://github.com/node-oauth/node-oauth2-server which also implements PKCE

**Reply by StorytellerCZ** [reply]
Awesome, let's schedule a meeting then to talk sort this out properly.

**Reply by jankapunkt** [reply]
Here are some OAuth2 resources we can use to prepare:

- https://datatracker.ietf.org/doc/html/rfc6749.html
- https://datatracker.ietf.org/doc/html/rfc6750.html
- https://datatracker.ietf.org/doc/html/rfc7636.html
- https://www.oauth.com/
